pkg/acsengine/munge.go

```diff
@@ -0,0 +1,194 @@
+func NormalizeForVMSSScaling(logger *logrus.Entry, templateMap map[string]interface{}) error {
+	if err := NormalizeMasterResourcesForScaling(logger, templateMap); err != nil {
+		return err
+	}
+
+	resources := templateMap[resourcesFieldName].([]interface{})
+	for _, resource := range resources {
+		resourceMap, ok := resource.(map[string]interface{})
+		if !ok {
+			logger.Warnf("Template improperly formatted")
+			continue
+		}
+
+		resourceType, ok := resourceMap[typeFieldName].(string)
+		if !ok || resourceType != vmssResourceType {
+			continue
+		}
+
+		resourceProperties, ok := resourceMap[propertiesFieldName].(map[string]interface{})
+		if !ok {
+			logger.Warnf("Template improperly formatted")
+			continue
+		}
+
+		virtualMachineProfile, ok := resourceProperties[virtualMachineProfileFieldName].(map[string]interface{})
+		if !ok {
+			logger.Warnf("Template improperly formatted")
+			continue
+		}
+
+		if !removeCustomData(logger, virtualMachineProfile) || !removeImageReference(logger, virtualMachineProfile) {
+			continue
+		}
+	}
+	return nil
+}
+
+// NormalizeForK8sVMASScalingUp takes a template and removes elements that are unwanted in a K8s VMAS scale up/down case
+func NormalizeForK8sVMASScalingUp(logger *logrus.Entry, templateMap map[string]interface{}) error {
+	if err := NormalizeMasterResourcesForScaling(logger, templateMap); err != nil {
+		return err
+	}
+	nsgIndex := -1
+	resources := templateMap[resourcesFieldName].([]interface{})
+	for index, resource := range resources {
+		resourceMap, ok := resource.(map[string]interface{})
+		if !ok {
+			logger.Warnf("Template improperly formatted")
+			continue
+		}
+
+		resourceType, ok := resourceMap[typeFieldName].(string)
+		if ok && resourceType == nsgResourceType {
+			if nsgIndex != -1 {
+				err := fmt.Errorf("Found 2 resources with type %s in the template. There should only be 1", nsgResourceType)
+				logger.Errorf(err.Error())
+				return err
+			}
+			nsgIndex = index
+		}
+
+		dependencies, ok := resourceMap[dependsOnFieldName].([]interface{})
+		if !ok {
+			logger.Warnf("Template improperly formatted")
+			continue
+		}
+
+		for dIndex := len(dependencies) - 1; dIndex >= 0; dIndex-- {
+			dependency := dependencies[dIndex].(string)
+			if strings.Contains(dependency, nsgResourceType) {
+				dependencies = append(dependencies[:dIndex], dependencies[dIndex+1:]...)
+			}
+		}
+
+		resourceMap[dependsOnFieldName] = dependencies
+	}
+	if nsgIndex == -1 {
+		err := fmt.Errorf("Found no resources with type %s in the template. There should have been 1", nsgResourceType)
+		logger.Errorf(err.Error())
+		return err
+	}
+
+	templateMap[resourcesFieldName] = append(resources[:nsgIndex], resources[nsgIndex+1:]...)
+
+	return nil
+}
+
+// NormalizeMasterResourcesForScaling takes a template and removes elements that are unwanted in any scale up/down case
+func NormalizeMasterResourcesForScaling(logger *logrus.Entry, templateMap map[string]interface{}) error {
```

Comment:
Hm, given my comment here, do you still think they need to be renamed? https://github.com/Azure/acs-engine/pull/615#discussion_r116417929

I could see this maybe:

`NormalizeForK8sVMASScalingUp ->`
`NormalizeForK8sVMASScaling` to be more consistent with:
`NormalizeForVMSSScaling` and changing this to be ->
`NormalizeForK8sVMSSScaling`

What do you think? Otherwise I feel yours are possibly inconsistent given spelling out "VirualMachine" but leaving VMSS.